include/drivers/rtc/ds3231.h

```diff
@@ -0,0 +1,640 @@
+	 * for alarm signals and will invoke the handler with a
+	 * parameter carrying the value returned by
+	 * rtc_ds3231_check_alarms().  The corresponding status flags
+	 * will be cleared in the device before the handler is
+	 * invoked.
+	 *
+	 * The handler will be invoked from the system work queue.
+	 */
+	rtc_ds3231_alarm_callback_handler_t handler;
+
+	/** @brief User-provided pointer passed to alarm callback. */
+	void *user_data;
+
+	/** @brief Flags controlling configuration of the alarm alarm.
+	 *
+	 * See RTC_DS3231_ALARM_FLAGS_IGNSE and related constants.
+	 *
+	 * Note that as described the alarm mask fields require that
+	 * if a unit is not ignored, higher-precision units must also
+	 * not be ignored.  For example, if match on hours is enabled,
+	 * match on minutes and seconds must also be enabled.  Failure
+	 * to comply with this requirement will cause
+	 * rtc_ds3231_set_alarm() to return an error, leaving the
+	 * alarm configuration unchanged.
+	 */
+	u8_t flags;
+};
+
+/** @brief Register the RTC clock against system clocks.
+ *
+ * This captures the same instant in both the RTC time scale and a
+ * stable system clock scale, allowing conversion between those
+ * scales.
+ */
+struct rtc_ds3231_syncpoint {
+	/** @brief Time from the DS3231.
+	 *
+	 * This maybe in UTC, TAI, or local offset depending on how
+	 * the RTC is maintained.
+	 */
+	struct timespec rtc;
+
+	/** @brief Value of a local clock at the same instant as #rtc.
+	 *
+	 * This is captured from a stable monotonic system clock
+	 * running at between 1 kHz and 1 MHz, allowing for
+	 * microsecond to millisecond accuracy in synchronization.
+	 */
+	u32_t syncclock;
+};
+
+typedef int (*rtc_ds3231_api_ctrl_update)(struct device *dev,
+					  u8_t set_bits,
+					  u8_t clear_bits);
+typedef int (*rtc_ds3231_api_stat_update)(struct device *dev,
+					  u8_t set_bits,
+					  u8_t clear_bits);
+
+typedef int (*rtc_ds3231_api_get_alarm)(struct device *dev,
+					u8_t idx,
+					struct rtc_ds3231_alarm *cfg);
+typedef int (*rtc_ds3231_api_set_alarm)(struct device *dev,
+					u8_t idx,
+					const struct rtc_ds3231_alarm *cfg);
+
+typedef int (*rtc_ds3231_api_synchronize) (struct device *dev,
+					   struct k_poll_signal *signal);
+typedef int (*rtc_ds3231_api_get_sync) (struct device *dev,
+					struct rtc_ds3231_syncpoint *syncpoint);
+typedef int (*rtc_ds3231_api_set) (struct device *dev,
+				   const struct rtc_ds3231_syncpoint *syncpoint,
+				   struct k_poll_signal *signal);
+
+typedef int (*rtc_ds3231_api_check_alarms) (struct device *dev);
+
+struct rtc_ds3231_driver_api {
+	struct counter_driver_api counter_api;
+	rtc_ds3231_api_ctrl_update ctrl_update;
+	rtc_ds3231_api_stat_update stat_update;
+	rtc_ds3231_api_get_alarm get_alarm;
+	rtc_ds3231_api_set_alarm set_alarm;
+	rtc_ds3231_api_synchronize synchronize;
+	rtc_ds3231_api_get_sync get_sync;
+	rtc_ds3231_api_set set;
+	rtc_ds3231_api_check_alarms check_alarms;
+};
+
```

Comment:
would be good to add build time assert which ensures that `counter_api` is the first item in the api struct.

Reply:
We don't generally do that in similar cases, like gpio which has required common prefixes.  It'd be pretty obvious if it was wrong.